### PR TITLE
Add Attachment structure

### PR DESCRIPTION
**Pull request by Lewdcario**
**Please describe the changes this PR makes and why it should be merged:**

Adding on to #1451,  I have added an attachment structure, which enables seamless integration of files into send parameters and into embed options, removing a lot of ugliness there once was when uploading even one file.

```js
const buffer = canvas.toBuffer();
// You can insert streams now.
const pngStream = canvas.createPNGStream();

message.channel.send('Buffer test', new Attachment(buffer, 'file.jpg'));
message.channel.send('PNGStream test', new Attachment(pngStream, 'file.png'));

// You can also provide a path.
message.channel.send(new Attachment('../dog.png'));

// You may provide a buffer without any name, which will be assumed to be 'file.jpg'.
message.channel.send(new Attachment(buffer));

// You may also provide an array of attachments
message.channel.send([new Attachment(buffer, 'file.jpg'), new Attachment(pngStream, 'file.png')]);

// In this case an error will be thrown as one of the attachments lacks a file and proper path.
message.channel.send([new Attachment(buffer, 'file.jpg'), new Attachment(pngStream, 'file.png'), new Attachment('some name')]);

// You can instantiate an embed in object literal notation or use some handy new methods.
let embed = new MessageEmbed().setTitle('Attachment').attachFiles([new Attachment(buffer, 'kek.jpg')]).setImage('attachment://kek.jpg');
embed = new MessageEmbed({ title: 'hey', files: [ new Attachment(buffer, 'kek.jpg') ], image: { url: 'attachment://kek.jpg' } });

// You can also send embeds directly now.
message.channel.send(new MessageEmbed().setTitle('eyy there'))

// Additionally, you may send both embed and files directly.
const files = [new Attachment(path.join(__dirname, '..', '..', 'assets', 'pngStream.png'))];
const embed = new MessageEmbed().setTitle(';)');
message.channel.send({ files, embed });

// Webhooks are now updated to be able to utilize all of the aforementioned features. Additionally, you can directly insert an array of embeds.
const hook = new WebhookClient('channelid', 'token');
const embed = new MessageEmbed().setDescription(':D').setColor('RED');
const files = [new Attachment('./assets/pngStream.png'), new Attachment(buffer, 'kek.jpg')];
hook.send(embed);
hook.send(new Attachment('./assets/pngStream.png'));
hook.send({ files, embed });
hook.send({ files, embeds: [embed, embed] });
```

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
- [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

File: src/client/ClientDataResolver.js [hunk 1]
```diff
@@ -216,6 +216,32 @@ class ClientDataResolver {
...
+        reject(new TypeError('REQ_RESOURCE_TYPE'));
+      } else {
+        this.resolveBuffer(resource)
+          .then(resolve)
+          .catch(() => {
+            if (resource.pipe && typeof resource.pipe === 'function') {
+              const buffers = [];
+              resource.on('error', reject);
+              resource.on('data', data => buffers.push(data));
+              resource.on('end', () => resolve(Buffer.concat(buffers)));
```
Comment (devsnek) [inline src/client/ClientDataResolver.js]: once

File: src/client/ClientDataResolver.js [hunk 2]
```diff
@@ -216,6 +216,32 @@ class ClientDataResolver {
...
+    return new Promise((resolve, reject) => {
+      if (!resource) {
+        reject(new TypeError('REQ_RESOURCE_TYPE'));
+      } else {
+        this.resolveBuffer(resource)
+          .then(resolve)
+          .catch(() => {
+            if (resource.pipe && typeof resource.pipe === 'function') {
+              const buffers = [];
+              resource.on('error', reject);
```
Comment (devsnek) [inline src/client/ClientDataResolver.js]: once

File: src/structures/interfaces/TextBasedChannel.js [hunk 3]
```diff
@@ -72,14 +74,24 @@ class TextBasedChannel {
    *  .then(message => console.log(`Sent message: ${message.content}`))
    *  .catch(console.error);
    */
-  send(content, options) {
+  send(content, options) { // eslint-disable-line complexity
     if (!options && typeof content === 'object' && !(content instanceof Array)) {
       options = content;
       content = '';
+      if (options instanceof MessageEmbed) options.embed = options;
```
Comment (devsnek) [inline src/structures/interfaces/TextBasedChannel.js]: move this outside the if statement

File: src/client/ClientDataResolver.js [hunk 4]
```diff
@@ -216,6 +216,32 @@ class ClientDataResolver {
...
+   * Converts a Stream to a Buffer.
+   * @param {Stream} resource The stream to convert
+   * @returns {Promise<Buffer>}
+   */
+  resolveFile(resource) {
+    return new Promise((resolve, reject) => {
+      if (!resource) {
+        reject(new TypeError('REQ_RESOURCE_TYPE'));
+      } else {
+        this.resolveBuffer(resource)
```
Comment (devsnek) [inline src/client/ClientDataResolver.js]: return this.resolveBuffer?

File: src/structures/Attachment.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
+const { resolveString } = require('../util/Util');
+const { TypeError } = require('../errors');
+
+/**
+ * Represents an attachment in a message
+ */
+class Attachment {
+  constructor(file, name) {
+    this.files = [];
```
Comment (devsnek) [inline src/structures/Attachment.js]: change to `this.file`, regarding conversation in staff channel

**Comment by 1Computer1** [comment]
Shouldn't the send method in Webhook.js also be updated to get these new features?

File: src/structures/Attachment.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    this._attach(file, name);
+  }
+
+  /**
+    * The name of the file
+    * @type {?string}
+    */
```
Comment (SpaceEEC) [inline src/structures/Attachment.js]: `@readonly`

File: src/structures/Attachment.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  get name() {
+    return this.files[0] ? this.files[0].name : null;
+  }
+
+  /**
+    * The file
+    * @type {?BufferResolvable|Stream}
+    */
```
Comment (SpaceEEC) [inline src/structures/Attachment.js]: `@readonly`

File: src/structures/interfaces/TextBasedChannel.js [hunk 8]
```diff
@@ -96,16 +108,18 @@ class TextBasedChannel {
             file.name = path.basename(file.attachment);
           } else if (file.attachment && file.attachment.path) {
             file.name = path.basename(file.attachment.path);
+          } else if (file instanceof Attachment) {
+            file = { name: path.basename(file.files[0]) || 'file.jpg', attachment: file.files[0] };
           } else {
             file.name = 'file.jpg';
           }
-        }
+        } else if (file instanceof Attachment) { file = file.files[0]; }
```
Comment (SpaceEEC) [inline src/structures/interfaces/TextBasedChannel.js]: Shouldn't this be across multiple lines?

**Pull request merged**